### PR TITLE
Add a cache for leader schedules

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Leader schedule is recomputed every time we ask for the leader for a slot. This will be an issue when we need constant access to the leader schedule in more demanding workloads, like verifying each received blob was sent by the correct leader for the slot.

#### Summary of Changes
Add a simple cache for leader schedules to be shared.

Fixes #

### CODE REVIEW

File: core/src/leader_schedule_utils.rs [hunk 1]
```diff
@@ -94,6 +78,22 @@ pub fn tick_height_to_slot(ticks_per_slot: u64, tick_height: u64) -> u64 {
     tick_height / ticks_per_slot
 }
 
+fn sort_stakes(stakes: &mut Vec<(Pubkey, u64)>) {
```
Comment (rob-solana) [inline core/src/leader_schedule_utils.rs]: is this function changed or just moved?  if just moved, how about not moving it ;)

File: core/src/leader_schedule_cache.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                let mut cached_schedules = self.cached_schedules.write().unwrap();
+                cached_schedules.insert(epoch, leader_schedule);
+                Self::retain_latest(&mut cached_schedules);
+                leader
+            })
+        }
+    }
+
+    fn retain_latest(schedules: &mut HashMap<u64, LeaderSchedule>) {
+        if schedules.len() > MAX_SCHEDULES {
```
Comment (rob-solana) [inline core/src/leader_schedule_cache.rs]: the Bank never culls epoch_vote_accounts, might not really be necessary for you to do any culling for a *long* time

File: core/src/leader_schedule_utils.rs (hunk 1 above)
Comment (carllin) [reply core/src/leader_schedule_utils.rs]: I want my private functions to come after my public functions :(

File: core/src/leader_schedule_cache.rs (hunk 2 above)
Comment (carllin) [reply core/src/leader_schedule_cache.rs]: 200 schedules? roughly a year?

Comment (rob-solana) [reply core/src/leader_schedule_cache.rs]: yeah, like noise in the scheme of things...

File: core/src/leader_schedule_utils.rs (hunk 1 above)
Comment (rob-solana) [reply core/src/leader_schedule_utils.rs]: ok, you go girl

Comment (rob-solana) [reply core/src/leader_schedule_utils.rs]: how about putting them in a separate impl block?

Comment (carllin) [reply core/src/leader_schedule_utils.rs]: hmmm there's no impl blocks in these, and a separate module seems like overkill

**Pull request merged**